src/Microsoft.Azure.WebJobs.Host/Dispatch/SharedQueueHandler.cs

```diff
@@ -0,0 +1,261 @@
+        {
+            if (_state != State.Started)
+            {
+                throw new InvalidOperationException(ErrorMessage(State.Started, _state));
+            }
+            // the count value does not change once SharedQueue is started, see "RegisterHandler()"
+            if (_triggerExecutor.HasMessageHandlers())
+            {
+                await _sharedQueuelistener.StopAsync(cancellationToken);
+            }
+            // if there's no messageHandlers registed, stopQueue is a NOOP
+            _state++;
+        }
+
+        private static string ErrorMessage(State expected, State actual)
+        {
+            return $"Expected state to be \"{expected}\" but actualy state is \"{actual}\", this is probably because methods are not called in correct order";
+        }
+
+        // initialize following fields async
+        // _triggerExecutor --> register messageHandler
+        // _sharedQueuelistener --> dequeue messages and call messageHandler
+        // _sharedQueueWriter --> enqueue messages
+        internal async Task InitializeAsync(CancellationToken cancellationToken)
+        {
+            if (_state != State.Created)
+            {
+                // only initialized once, since _state is incremental
+                throw new InvalidOperationException(ErrorMessage(State.Created, _state));
+            }
+
+            // concurrent dictionary that we can register messageHandler
+            _triggerExecutor = new SharedQueueExecutor();
+
+            try
+            {
+                IStorageQueueClient primaryQueueClient = (await _accountProvider.GetStorageAccountAsync(cancellationToken)).CreateQueueClient();
+                string hostId = await _hostIdProvider.GetHostIdAsync(cancellationToken);
+
+                // one host level shared queue
+                // queue is not created here, only after 1st message added
+                IStorageQueue sharedQueue = primaryQueueClient.GetQueueReference(HostQueueNames.GetHostSharedQueueName(hostId));
+                // default host level poison queue
+                IStorageQueue sharedPoisonQueue = primaryQueueClient.GetQueueReference(HostQueueNames.GetHostSharedPoisonQueueName(hostId));
+
+                // queueWatcher will update queueListener's polling interval when queueWriter performes an enqueue operation
+                SharedQueueWatcher sharedQueueWatcher = _sharedContextProvider.GetOrCreateInstance<SharedQueueWatcher>(
+                                                            new SharedQueueWatcherFactory(_messageEnqueuedWatcherSetter));
+                _sharedQueueWriter = new SharedQueueWriter(sharedQueue, sharedQueueWatcher);
+
+                // use default poisonQueue setup
+                _sharedQueuelistener = new QueueListener(sharedQueue, sharedPoisonQueue, _triggerExecutor,
+                    _exceptionHandler, _trace, _loggerFactory, sharedQueueWatcher, _queueConfiguration);
```

Comment:
So if there's no storage (AFR), then all this gets skipped. Is that ok ... will it cause problems later? For example, ListenerFactoryListener  looks like it calls Initialize() and then StartQueueAsync(); but even though Initialize() returns "Safely", StartQueueAsync() which will just throw.

Reply:
I'd suggest having an explicit State, like "unavailable", and set that.  

---
In reply to: [147010623](https://github.com/Azure/azure-webjobs-sdk/pull/1362#discussion_r147010623) [](ancestors = 147010623)